solutions/usaco-941.mdx

```diff
@@ -11,11 +11,161 @@ author: Jesse Choe, Ryan Chou
+
+It will help to actually look at things from the viewpoint of the characteristics instead of from the viewpoint of the populations, so let's "transpose" the input above:
+
+$\texttt{A: population1 population3 } \\
+\texttt{B: population2 population3}$
+
+The fundamental problem here is that there are populations in $\texttt{A}$ only, populations in $\texttt{B}$ only, and populations in both $\texttt{A}$ and $\texttt{B}$. If we look at the Venn diagram for the sets A and B, the picture therefore looks like this:
+
+![Venn Diagram 1](http://www.usaco.org/current/data/fig2_evolution_bronze_open19.png)
+
+Let's call this situation a "crossing" pair of sets. In general, two sets can be disjoint (no overlap), nesting (one inside the other), or crossing (overlap but not nesting). If any two of the characteristics A and B in our instance represent crossing sets as above, then we cannot build a proper tree. On the other hand, if all the characteristics represent sets that don't cross (they are either disjoint or nested), then we get a Venn diagram like this:
+
+![Venn Diagram 2](http://www.usaco.org/current/data/fig3_evolution_bronze_open19.png)
+
+If you look at this picture carefully, hopefully you see a tree formed by the nesting structure of the sets:
+
+![Evolution tree diagram](http://www.usaco.org/current/data/fig4_evolution_bronze_open19.png)
+
+A tree like this is easy to convert into a proper evolutionary tree. E.g., if we have three children A, B, and C, we could just make three sequential two-way splits that add the A, B and C characteristics.
+
+So, we actually don't need to build a proper evolutionary tree, but we just need to test if any of our characteristics represent crossing sets; if so (and only if so), a proper tree is impossible to build. This leads to the solution where we build all the sets of populations having each characteristic and then test if any pair of these sets is crossing.
+
 ## Implementation
 
 **Time Complexity:** $\mathcal{O}(N^2)$
 
 <LanguageSection>
+
+<CPPSection>
+
+```cpp
+// Taken from official analysis by Brian Dean
+#include <iostream>
+#include <fstream>
+#include <vector>
+using namespace std;
+ 
+int N;
+vector<string> characteristics[25];
+vector<string> all_characteristics;
+ 
+// Do two sets "cross" -- I.e., are there elements in A, B, and A intersect B?
+bool crossing(int a, int b)
+{
+	int A=0, B=0, AB=0;
+	for (int i=0; i<N; i++) {
+	vector<string> &v = characteristics[i];
+	bool has_a = false, has_b = false;
+	for (int j=0; j<v.size(); j++) {
+	  if (v[j]==all_characteristics[a]) has_a = true;
+	  if (v[j]==all_characteristics[b]) has_b = true;
```

Comment:
add spaces around operators

same for the rest of the code

and the java code too